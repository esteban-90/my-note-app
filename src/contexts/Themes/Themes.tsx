import type { FC, PropsWithChildren } from 'react'
import type { Theme } from '@emotion/react'
import { ThemeProvider as Provider } from '@emotion/react'
import useLocalStorage from 'use-local-storage'
import { Themes } from '@/styles'

const defaultValue: Theme['themeName'] = matchMedia('prefers-color-scheme: dark').matches ? 'night' : 'day'

/**
 * Component for providing theme to app.
 * @returns The provider.
 */

export const ThemeProvider: FC<PropsWithChildren> = ({ children }) => {
  const [savedTheme, setSavedTheme] = useLocalStorage('theme', defaultValue)
  const currentTheme = Themes[savedTheme]

  const toggleTheme = () => {
    setSavedTheme((previousTheme) => (previousTheme === 'night' ? 'day' : 'night'))
  }

  const value: Theme = {
    themeName: currentTheme.themeName,
    colorPalette: currentTheme.colorPalette,
    toggleTheme,
  }

  return <Provider theme={value}>{children}</Provider>
}
