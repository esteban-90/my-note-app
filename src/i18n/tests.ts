import type { Resource } from 'i18next'
import { createInstance } from 'i18next'
import { initReactI18next } from 'react-i18next'
import { ns, supportedLngs } from './app'

const resources: Resource = {}
export const i18nInstance = createInstance()

void (async () => {
  for (const n of ns) {
    for (const lng of supportedLngs) {
      resources[lng] ??= {}
      resources[lng] = {
        ...resources[lng],
        [n]: await import(`../../public/translations/${lng}/${n}.json`),
      }
    }
  }
})()

i18nInstance.use(initReactI18next).init({
  ns,
  supportedLngs,
  fallbackLng: supportedLngs[0],
  lng: supportedLngs[0],
  react: { useSuspense: false },
  resources,
})
