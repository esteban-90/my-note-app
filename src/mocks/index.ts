import type { Theme } from '@emotion/react'
import type { Note } from '@/types'
import uniqid from 'uniqid'

let seconds = 0

export const makeMockNote = (content: string): Note => {
  return {
    id: uniqid('note-'),
    content,
    createdAt: new Date(1995, 11, 17, 3, 24, seconds++).toISOString(),
  }
}

const makeMockNotes = (quantity: number) => {
  return new Array(quantity).fill(null).map((_, index) => makeMockNote(`Hello test ${index + 1}\nTesting ${index + 1}`))
}

export const mockNote = makeMockNote('Testing\nNote')
export const mockNotes = makeMockNotes(5)

const noteKey = 'notes'
const themeKey = 'theme'

export const getNotes = () => {
  const savedNotes = localStorage.getItem(noteKey) as string
  const parsedNotes = JSON.parse(savedNotes) as Note[]
  return parsedNotes
}

export const saveNotes = () => {
  localStorage.setItem(noteKey, JSON.stringify(mockNotes))
}

export const removeNotes = () => {
  localStorage.removeItem(noteKey)
}

export const getTheme = () => {
  const savedTheme = localStorage.getItem(themeKey) as Theme['themeName']
  return savedTheme
}

export const saveTheme = () => {
  const nightTheme: Theme['themeName'] = 'night'
  localStorage.setItem(themeKey, JSON.stringify(nightTheme))
}

export const removeTheme = () => {
  localStorage.removeItem(themeKey)
}
