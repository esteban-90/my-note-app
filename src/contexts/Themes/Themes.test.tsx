import { useTheme } from '@emotion/react'
import { getTheme, removeTheme } from '@/mocks'
import { screen, render, userEvent, configure } from '@/tests'
import { ThemeProvider } from './Themes'

describe(ThemeProvider.name, () => {
  let asFragment: () => DocumentFragment

  const Child = () => {
    const { themeName, toggleTheme } = useTheme()

    return (
      <>
        <p>{themeName}</p>
        <button onClick={toggleTheme}>Change theme</button>
      </>
    )
  }

  const UI = (
    <ThemeProvider>
      <Child />
    </ThemeProvider>
  )

  beforeAll(() => {
    configure({ throwSuggestions: true })
  })

  beforeEach(() => {
    ;({ asFragment } = render(UI))
  })

  it('should match snapshot', () => {
    expect(asFragment()).toMatchSnapshot()
  })

  it('should provide a specific theme at a time to child elements and save it to storage', () => {
    const themedElement = screen.getByText('night')
    expect(themedElement).toBeInTheDocument()

    const savedTheme = getTheme()
    expect(savedTheme).not.toBeNull()
    expect(savedTheme).not.toContain('day')
    expect(savedTheme).toContain('night')
  })

  it('should be able to change theme and save it to storage with the new value', async () => {
    const button = screen.getByRole('button', { name: /^change theme$/i })
    await userEvent.click(button)

    const themedElement = await screen.findByText('day')
    expect(themedElement).toBeInTheDocument()

    const savedTheme = getTheme()
    expect(savedTheme).not.toBeNull()
    expect(savedTheme).not.toContain('night')
    expect(savedTheme).toContain('day')

    removeTheme()
  })
})
