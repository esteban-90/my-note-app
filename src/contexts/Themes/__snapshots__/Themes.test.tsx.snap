// Vitest Snapshot v1

exports[`ThemeProvider > should match snapshot 1`] = `
<DocumentFragment>
  <p>
    night
  </p>
  <button>
    Change theme
  </button>
</DocumentFragment>
`;
